### PR TITLE
BUG: Fix parse_dates processing with usecols and C engine

### LINKED ISSUE #9755: read_csv with names, usecols and parse_dates
xref #12203 

The arrays passed to the date_parser function is different when `names` and `use_cols` are specified to limit the number of parsed columns.

When running the example code below, the date_parser function receives two arguments, one array with '20140101' strings, and one array with integers. The default `date_parser` fails to process this input.

When assigning an empty list to `DROPPED_COLUMNS` (so that all columns are parsed), the second array contains strings instead of integers, and the datetimes are parsed correctly.

The problem doesn't occur with `engine='python'`. I haven't tested the influence of the `header` and `index_cols` options.

Python script:

	from __future__ import print_function, division
	import pandas as pd

	CSV = '2014.csv'

	DROPPED_COLUMNS = ['NCDC', 'I', 'QCP']
	# DROPPED_COLUMNS = []

	with open(CSV) as csv:
	    csv.readline()
	    column_names = csv.readline().split()
	    used_columns = [i for i, column_name in enumerate(column_names)
	                    if column_name not in DROPPED_COLUMNS]
	    used_col_names = [column_name for i, column_name in enumerate(column_names)
	                      if i in used_columns]
	    parse_dates = [[i for i, column_name in enumerate(used_col_names)
	                    if column_name in ('Date', 'HrMn')]]
	    print(parse_dates)

	    data = pd.read_csv(csv, header=None, names=used_col_names, index_col=False, engine='python',
	                       parse_dates=parse_dates, usecols=used_columns)
	    print(data)

Contents of 2014.csv:

	Identification                          TEMP
	USAF   NCDC  Date     HrMn I Type  QCP  Temp   Q
	062693,99999,20140101,0025,4,FM-15,    ,   7.0,1,
	062693,99999,20140101,0055,4,FM-15,    ,   6.0,1,
	062693,99999,20140101,0125,4,FM-15,    ,   6.0,1,
	062693,99999,20140101,0155,4,FM-15,    ,   6.0,1,
	062693,99999,20140101,0225,4,FM-15,    ,   6.0,1,
	062693,99999,20140101,0255,4,FM-15,    ,   6.0,1,
	062693,99999,20140101,0325,4,FM-15,    ,   6.0,1,
	062693,99999,20140101,0355,4,FM-15,    ,   6.0,1,

### LINKED ISSUE #12678: ERR: dont' allow ambiguous usecols
https://github.com/pydata/pandas/pull/12551 and comment: https://github.com/pydata/pandas/pull/12512#discussion_r56771668

``usecols=['a',1]`` should raise

### PR DESCRIPTION
**Pull request by gfyoung**
closes #9755 
closes #12678 

Continuing on my conquest of `read_csv` bugs, this PR fixes a bug brought up in #9755 in processing `parse_dates` with the C engine in which the wrong indices (those of the filtered column names) were being used to determine the date columns to not be dtype-parsed by the C engine. The correct indices are those of the original column names, as they are used later on in the actual data processing.

### CODE REVIEW

**Comment by gfyoung** [comment]
@jreback : Tests are passing once more, so this should be good to merge AFAICT.

File: pandas/io/parsers.py [hunk 1]
```diff
@@ -1179,13 +1189,26 @@ def __init__(self, src, **kwds):
...
+        # 'names' and 'orig_names' will either
+        # be identical, or 'names' will be a
+        # strict subset of 'orig_names'
+        identical = (len(self.names) == len(self.orig_names))
+        names = self.orig_names
 
         def _set(x):
-            if com.is_integer(x):
-                self._reader.set_noconvert(x)
+            if not identical:
```
Comment (jreback) [inline pandas/io/parsers.py]: seems pretty duplicative here. pls try to make minimal code changes.

File: doc/source/whatsnew/v0.18.0.txt [hunk 2]
```diff
@@ -1201,3 +1201,4 @@ Bug Fixes
 - Bug when initializing categorical series with a scalar value. (:issue:`12336`)
 - Bug when specifying a UTC ``DatetimeIndex`` by setting ``utc=True`` in ``.to_datetime`` (:issue:`11934`)
 - Bug when increasing the buffer size of CSV reader in ``read_csv`` (:issue:`12494`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.18.0.txt]: this would be for 0.18.1

Comment (gfyoung) [inline doc/source/whatsnew/v0.18.0.txt]: Done.

File: pandas/io/parsers.py (hunk 1 above)
Comment (gfyoung) [inline pandas/io/parsers.py]: Done.

**Comment by gfyoung** [comment]
@jreback : Tests are passing once more, so this should be good to merge AFAICT.

**Comment by gfyoung** [comment]
@jreback : Tests are passing once more, so this should be good to merge AFAICT.

File: pandas/io/parsers.py [hunk 3]
```diff
@@ -1183,13 +1193,22 @@ def __init__(self, src, **kwds):
         self._implicit_index = self._reader.leading_cols > 0
 
     def _set_noconvert_columns(self):
-        names = self.names
+        # By the time this function is called,
+        # 'names' and 'orig_names' will either
+        # be identical, or 'names' will be a
+        # strict subset of 'orig_names'
+        identical = (len(self.names) == len(self.orig_names))
```
Comment (jreback) [inline pandas/io/parsers.py]: can these ever be different? when are they not identical?

Comment (gfyoung) [inline pandas/io/parsers.py]: They are not identical when you specify `usecols`.  If you see where this method is called, it is called right after `self.names` is filtered by `usecols`.  The fix is I set `self.orig_names` to `self.names` before filtering so that we have a copy of the original columns before filtering.  The fact we didn't have a copy of the unfiltered column list was why this bug emerged.

File: doc/source/whatsnew/v0.18.1.txt [hunk 4]
```diff
@@ -138,3 +138,4 @@ Bug Fixes
 
 
 - Bug in ``pivot_table`` when ``margins=True`` and ``dropna=True`` where nulls still contributed to margin count (:issue:`12577`)
+- Bug when specifying `usecols` and `parse_dates` simultaneously with the C engine (:issue:`9755`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.18.1.txt]: double back-ticks here on these args, say in ``.read_csv()``

**Comment by jreback** [comment]
small doc changes. looks good otherwise.

**Comment by gfyoung** [comment]
Sounds good.  I'll let Travis do its thing, and if tests pass, I'll make the doc change and ping you.

**Comment by gfyoung** [comment]
@jreback : Travis is passing, and the doc change has been made.  Should be good to merge if there is nothing else.

**Comment by gfyoung** [comment]
@jreback : Travis is passing, and the doc change has been made.  Should be good to merge if there is nothing else.

**Comment by gfyoung** [comment]
@jreback : Travis is passing, and the doc change has been made.  Should be good to merge if there is nothing else.

**Comment by jreback** [comment]
the first part of the example still fails. your tests cover the 2nd part (which we *thought* were the same).

```
s = """0,1,20140101,0900,4
0,1,20140102,1000,4"""

# using all cols -> no problem
pd.read_csv(StringIO(s), header=None, names=list('abcde'), parse_dates=[[2,3]], engine='c')

# using certain cols -> problem
pd.read_csv(StringIO(s), header=None, names=list('acd'), usecols=[0,2,3], parse_dates=[[1,2]], engine='c')

# but not with python engine
pd.read_csv(StringIO(s), header=None, names=list('acd'), usecols=[0,2,3], parse_dates=[[1,2]], engine='python')
```

**Comment by gfyoung** [comment]
Ah, okay.  Did not realize that.  I'll take a look and see how I can fix that.

**Comment by gfyoung** [comment]
@jreback : Travis has confirmed that the bug has now been squashed

File: pandas/io/parsers.py [hunk 5]
```diff
@@ -1146,18 +1146,27 @@ def __init__(self, src, **kwds):
...
+        #
+        # once names has been filtered, we will
+        # then set orig_names again to names
+        self.orig_names = self.names[:]
+
+        if self.usecols:
+            if len(self.names) > len(self.usecols):
+                col_indices = []
+                for u in self.usecols:
+                    if isinstance(u, string_types):
```
Comment (jreback) [inline pandas/io/parsers.py]: you don't need all this, something like:
```
indexer = Index(self.usecols).get_indexer_for(self.names)
indexer = indexer[indexer!=-1]
self.names = Index(self.names).take(indexer)
```

handles duplicates as well

File: pandas/io/parsers.py [hunk 6]
```diff
@@ -1183,13 +1192,30 @@ def __init__(self, src, **kwds):
...
     def _set_noconvert_columns(self):
-        names = self.names
+        # By the time this function is called, 'names' and
+        # 'orig_names' will either be identical, or 'names'
+        # will be a strict subset of 'orig_names'
+        same_as_orig = (len(self.names) == len(self.orig_names))
+
+        # If the original names are for the filtered table, then
+        # 'usecols' is our only resource in providing the correct
+        # column indices for the original table
```
Comment (jreback) [inline pandas/io/parsers.py]: see if you can make this simpler / less special casey.

File: pandas/io/parsers.py (hunk 5 above)
Comment (gfyoung) [inline pandas/io/parsers.py]: Everything broke when I did that.  If `usecols` is just a list of integers, I just get a string of `-1`'s for the `indexer` variable, so nothing `names` is just an empty array.

Comment (gfyoung) [inline pandas/io/parsers.py]: I would also add that this is not new machinery I've added but rather a refactoring.  The current implementation also takes into account the fact you could have a mixture of column names and indices for `usecols`.

Comment (jreback) [inline pandas/io/parsers.py]: @gfyoung obviously. nothing much in pandas is actually *new*. However, pls use the idiomatic code. Loops just require you to reinvent the wheel.

You need to be sure that you are ONLY working with strings.
```
In [1]: Index([1,2,3]).get_indexer_for([1,1])
Out[1]: array([0, 0])

In [2]: Index([1,2,3]).get_indexer_for(['1','1'])
Out[2]: array([-1, -1])

```

File: pandas/io/parsers.py (hunk 6 above)
Comment (gfyoung) [inline pandas/io/parsers.py]: Ah, there was some overlap in the logic.  Done.

File: pandas/io/parsers.py (hunk 5 above)
Comment (gfyoung) [inline pandas/io/parsers.py]: How can I check without using a loop that every element in `self.usecols` is a string?

Comment (jreback) [inline pandas/io/parsers.py]:
```
In [7]: Index([1,2,3]).astype(str)
Out[7]: Index([u'1', u'2', u'3'], dtype='object')
```

Comment (gfyoung) [inline pandas/io/parsers.py]: Okay, let me clarify my question.  Given that your columns are `['a', 'b', 'c', 'd', 'e']` and the user passes in `usecols=['b', 3, 4]`, how do I enforce that every element in `self.usecols` is a string?

Comment (gfyoung) [inline pandas/io/parsers.py]: Actually, if you want to simplify it, here's a way:
~~~python
names = [n for i, n in enumerate(self.names)
         if (i in self.usecols or n in self.usecols)]
self.names = list(set(names))

~~~

Comment (jreback) [inline pandas/io/parsers.py]: that will should fail, how are you supposed to interpret a mixed ``usecols``? what if you columns are: ``['3', '4', 'b', '1', '2']``.

If don't think we have restrictions on ``usecols`` ATM, but they should either be all positional (e.g. integer dtype), or not.

In fact, pls you put an ambiguous ``usecols`` in your last PR: https://github.com/pydata/pandas/pull/12551.

pls fix that. So we need validation on ``usecols``. 

you can do something like:

```
In [8]: pd.lib.infer_dtype([1,2,3])
Out[8]: 'integer'

In [9]: pd.lib.infer_dtype(['1','2','3'])
Out[9]: 'string'

In [10]: pd.lib.infer_dtype(['1','2',3])
Out[10]: 'mixed-integer'
```

and raise on [10])

Comment (jreback) [inline pandas/io/parsers.py]: https://github.com/pydata/pandas/issues/12678

Comment (gfyoung) [inline pandas/io/parsers.py]: Ah...then that's a hole in documentation because that isn't mentioned anywhere, and as you saw, the ones in #12551 pass just fine on both the Python and C ends.

Comment (gfyoung) [inline pandas/io/parsers.py]: I decided to go with the one-liner I suggested earlier because your suggestion unfortunately is actually an API change.  In the original loop, the order of the columns in the original data is preserved _independent_ of the order of `usecols`, whereas `take(indexer)` makes it _dependent_.  This causes several test failures like this one <a href="https://github.com/pydata/pandas/blob/master/pandas/io/tests/test_parsers.py#L2243">here</a>.

**Comment by gfyoung** [comment]
@jreback : Sorry, accidentally pushed a little earlier during the changes.  Could you cancel my earlier build <a href="https://travis-ci.org/pydata/pandas/builds/117345807">here</a>.  Thanks!

**Comment by gfyoung** [comment]
@jreback : Simplified the logic, and the tests are passing.  Should be good to merge I think.

**Comment by gfyoung** [comment]
@jreback : Simplified the logic, and the tests are passing.  Should be good to merge I think.

File: pandas/io/parsers.py [hunk 7]
```diff
@@ -1146,18 +1146,21 @@ def __init__(self, src, **kwds):
...
+        # so that proper indexing can be done
+        # with _set_noconvert_columns
+        #
+        # once names has been filtered, we will
+        # then set orig_names again to names
+        self.orig_names = self.names[:]
+
+        if self.usecols:
+            if len(self.names) > len(self.usecols):
+                self.names = [n for i, n in enumerate(self.names)
```
Comment (jreback) [inline pandas/io/parsers.py]: this doesn't handle the mixed usecols issue.

Comment (gfyoung) [inline pandas/io/parsers.py]: From an organization perspective, I was thinking of fixing usecols issue as a follow-up because it's a problem for both the Python and C engines, probably by putting a verification step in `ParserBase`.  Or should I just kill two birds with one stone and fix it here (+ tests)?

Comment (jreback) [inline pandas/io/parsers.py]: Well we have lots of things floating ATM. However I think its simple to add the non-mixed usecols (w/o addressing the set/list issue). So you can do that in another commit if it works.

Comment (gfyoung) [inline pandas/io/parsers.py]: Sound good.  I'll just add it as another commit and then close the issue in the process.

**Comment by gfyoung** [comment]
@jreback : Added the fix in for mixed-dtype columns.  Travis has given the green light, so this is ready to merge if there is nothing else to change / add.

File: doc/source/whatsnew/v0.18.1.txt [hunk 8]
```diff
@@ -164,3 +164,9 @@ Bug Fixes
 
 
 - Bug in ``pivot_table`` when ``margins=True`` and ``dropna=True`` where nulls still contributed to margin count (:issue:`12577`)
+- Bug in ``read_csv`` when specifying ``usecols`` and ``parse_dates`` simultaneously with the C engine (:issue:`9755`)
+
+
+
+
+- ``read_csv`` no allows a combination and strings and integers for the ``usecols`` parameter (:issue:`12678`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.18.1.txt]: put this in API changes (and it 'no longer allows').

File: pandas/io/parsers.py [hunk 9]
```diff
@@ -75,8 +75,11 @@ class ParserWarning(Warning):
     of each line, you might consider index_col=False to force pandas to _not_
     use the first column as the index (row names)
 usecols : array-like, default None
-    Return a subset of the columns.
-    Results in much faster parsing time and lower memory usage.
+    Return a subset of the columns. All elements in this array must either
+    be positional (i.e. integer indices into the document columns) or strings
```
Comment (jreback) [inline pandas/io/parsers.py]: maybe add in a mini example: ``[0, 1, 2]``, ``['foo', 'bar', 'baz']``.

File: pandas/io/parsers.py [hunk 10]
```diff
@@ -1146,18 +1158,21 @@ def __init__(self, src, **kwds):
...
+        # need to set orig_names here first
+        # so that proper indexing can be done
+        # with _set_noconvert_columns
+        #
+        # once names has been filtered, we will
+        # then set orig_names again to names
+        self.orig_names = self.names[:]
+
+        if self.usecols:
+            if len(self.names) > len(self.usecols):
```
Comment (jreback) [inline pandas/io/parsers.py]: I don't think this check is really valid. You might have len(names) < len(usecols) (which is kind of silly, but possible).
I think if ``.usecols`` is specified you *always* need to hit this path. Maybe try to find a test which breaks this.

File: pandas/io/parsers.py [hunk 11]
```diff
@@ -1146,18 +1158,21 @@ def __init__(self, src, **kwds):
...
+        #
+        # need to set orig_names here first
+        # so that proper indexing can be done
+        # with _set_noconvert_columns
+        #
+        # once names has been filtered, we will
+        # then set orig_names again to names
+        self.orig_names = self.names[:]
+
+        if self.usecols:
```
Comment (jreback) [inline pandas/io/parsers.py]: also at this point you *know* if you have positional usecols or named uscols. so use that information. I don't think you need to whole ``_set`` thing (on a per-name basis), you already know what to do.

File: pandas/io/parsers.py [hunk 12]
```diff
@@ -790,6 +793,14 @@ def _is_index_col(col):
     return col is not None and col is not False
 
 
+def _validate_usecols_arg(usecols):
```
Comment (jreback) [inline pandas/io/parsers.py]: then return the inferred type (which you can use later on).

File: doc/source/whatsnew/v0.18.1.txt (hunk 8 above)
Comment (gfyoung) [inline doc/source/whatsnew/v0.18.1.txt]: Good catch.   Done.

File: pandas/io/parsers.py (hunk 9 above)
Comment (gfyoung) [inline pandas/io/parsers.py]: Done.

File: pandas/io/parsers.py (hunk 10 above)
Comment (gfyoung) [inline pandas/io/parsers.py]: I'm not sure I follow your comment.  If you have `len(names) < len(usecols)`, you would be automatically erroring just as we do already.

Comment (jreback) [inline pandas/io/parsers.py]: if that is the case then the check is not necessary? can you confirm and add a test (or maybe move these test next to the other ones).

File: pandas/io/parsers.py (hunk 12 above)
Comment (gfyoung) [inline pandas/io/parsers.py]: Okay, I see.  Done.

File: pandas/io/parsers.py (hunk 10 above)
Comment (gfyoung) [inline pandas/io/parsers.py]: The check is necessary because if you have a situation where the user deliberately passes in only the column names for the filtered table but `usecols` corresponding to the indices of the original table, you are going to get an error.

File: pandas/io/parsers.py (hunk 11 above)
Comment (gfyoung) [inline pandas/io/parsers.py]: The reason why you have to set on a per-name basis is because `parse_dates` can be a mixture of strings and positional integers.  Knowing whether `usecols` is `string`, `integer`, or `None` (i.e. when `usecols` is not specified) does not help much here.

**Comment by gfyoung** [comment]
@jreback : Any updates with respect to my responses to your refactoring and check comments?

File: doc/source/whatsnew/v0.18.1.txt [hunk 13]
```diff
@@ -236,3 +240,12 @@ Bug Fixes
 - Bug in ``.describe()`` resets categorical columns information (:issue:`11558`)
 - Bug where ``loffset`` argument was not applied when calling ``resample().count()`` on a timeseries (:issue:`12725`)
 - ``pd.read_excel()`` now accepts path objects (e.g. ``pathlib.Path``, ``py.path.local``) for the file path, in line with other ``read_*`` functions (:issue:`12655`)
+
+
+
+
+
+- Bug in ``read_csv`` when specifying ``usecols`` and ``parse_dates`` simultaneously with the C engine (:issue:`9755`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.18.1.txt]: don't add onto the end, put in-between other bug fixes (that way you won't have, nor cause conflicts)

File: pandas/io/parsers.py [hunk 14]
```diff
@@ -75,8 +75,12 @@ class ParserWarning(Warning):
     of each line, you might consider index_col=False to force pandas to _not_
     use the first column as the index (row names)
 usecols : array-like, default None
-    Return a subset of the columns.
```
Comment (jreback) [inline pandas/io/parsers.py]: make sure you update ``io.rst`` with the same

File: pandas/io/tests/test_parsers.py [hunk 15]
```diff
@@ -2682,14 +2682,100 @@ def test_uneven_lines_with_usecols(self):
...
+        tm.assert_frame_equal(df, expected)
+
+        df = read_csv(StringIO(s), names=names,
+                      usecols=[3, 0, 2],
+                      parse_dates=parse_dates)
+        tm.assert_frame_equal(df, expected)
+
+    def test_mixed_dtype_usecols(self):
+        # See gh-12678
+        data = """a,b,c
```
Comment (jreback) [inline pandas/io/tests/test_parsers.py]: I think its worth testing with a header that is say ``[2, 0, 1]``, and usecols is specified as ``[0, 2]``. IOW, the header are actually labels AND we are specifying labels, but these are actually numbers, so it *looks* positional.

File: doc/source/whatsnew/v0.18.1.txt (hunk 13 above)
Comment (gfyoung) [inline doc/source/whatsnew/v0.18.1.txt]: Done.

File: pandas/io/parsers.py (hunk 14 above)
Comment (gfyoung) [inline pandas/io/parsers.py]: Ah, good point.  Done.

File: pandas/io/tests/test_parsers.py (hunk 15 above)
Comment (gfyoung) [inline pandas/io/tests/test_parsers.py]: 1) I'm not quite I sure I understand what you're saying, is it something like this?
~~~python
>>> data = """a,b,c
1000,2000,3000
4000,5000,6000
"""
>>> usecols = [0, 2] # column selection by index
>>> read_csv(StringIO(data), usecols=usecols)
>>>
>>> usecols = ['0', '2'] # column selection by name
>>> read_csv(StringIO(data), usecols=usecols)
~~~
2) Where or how does this relate to the mixed-dtypes?

Comment (jreback) [inline pandas/io/tests/test_parsers.py]: In [3]: read_csv(StringIO(data))
Out[3]: 
   2  3  4
0  a  b  c
1  d  e  f

In [5]: read_csv(StringIO(data),usecols=[2])
Out[5]: 
   4
0  c
1  f
```

Comment (jreback) [inline pandas/io/tests/test_parsers.py]: this is clearly doing positional indexing, even though the label exists. not sure what to do here.

Comment (gfyoung) [inline pandas/io/tests/test_parsers.py]: As is stated in the new documentation, integers are interpreted as indices.  Strings are interpreted as column names.  Personally, I consider that behaviour to be correct.

Comment (jreback) [inline pandas/io/tests/test_parsers.py]: hmm, ok

Comment (gfyoung) [inline pandas/io/tests/test_parsers.py]: I'll go ahead and add a separate test in there just in case this confusion comes up later on.

File: pandas/io/tests/test_parsers.py [hunk 16]
```diff
@@ -2682,14 +2682,100 @@ def test_uneven_lines_with_usecols(self):
...
+        cols = {
+            'a'  : [0, 0],
+            'c_d': [
+                Timestamp('2014-01-01 09:00:00'),
+                Timestamp('2014-01-02 10:00:00')
+            ]
+        }
+        expected = DataFrame(cols, columns=['c_d', 'a'])
+
+        df = read_csv(StringIO(s), usecols=[0, 2, 3],
```
Comment (jreback) [inline pandas/io/tests/test_parsers.py]: you are not testing this fully here. you MUST use ``self.read_csv``. In fact we should REMOVE ``read_csv`` from the import as its just confusing.

Comment (jreback) [inline pandas/io/tests/test_parsers.py]: ``read_csv`` will default to the c-engine ONLY. while ``self.read_csv`` cycles thru all of them (as separate tests)

File: pandas/io/tests/test_parsers.py (hunk 15 above)
Comment (jreback) [inline pandas/io/tests/test_parsers.py]: ok, just document it

File: pandas/io/tests/test_parsers.py (hunk 16 above)
Comment (gfyoung) [inline pandas/io/tests/test_parsers.py]: Whoops, let me go fix that.

File: pandas/io/tests/test_parsers.py (hunk 15 above)
Comment (gfyoung) [inline pandas/io/tests/test_parsers.py]: @jreback : Double checked on the documentation to make sure that the wording clearly states that integers = indices and strings = column names.  Also added a test where column names were integers.  Travis gives the green light.

File: pandas/io/parsers.py [hunk 17]
```diff
@@ -1133,6 +1151,7 @@ def __init__(self, src, **kwds):
 
         # XXX
         self.usecols = self._reader.usecols
+        self._usecols_dtype = _validate_usecols_arg(self.usecols)
```
Comment (jreback) [inline pandas/io/parsers.py]: I don't think you are actually using this variable? ``_usecols_dtype``, so no need to save it (I think we discussed using it, but not necessary i  guess)

File: pandas/io/parsers.py [hunk 18]
```diff
@@ -801,6 +805,20 @@ def _is_index_col(col):
...
+    usecols_dtype = None
+
+    if usecols is not None:
+        usecols_dtype = lib.infer_dtype(usecols)
+        if usecols_dtype not in ('integer', 'string'):
+            raise ValueError(("The elements of 'usecols' "
+                              "must either be all strings "
+                              "or all integers"))
+
+    return usecols_dtype
```
Comment (jreback) [inline pandas/io/parsers.py]: no need to have a return

File: pandas/io/parsers.py [hunk 19]
```diff
@@ -801,6 +805,20 @@ def _is_index_col(col):
     return col is not None and col is not False
 
 
+def _validate_usecols_arg(usecols):
+    # gh-12678
+    usecols_dtype = None
```
Comment (jreback) [inline pandas/io/parsers.py]: remove this

File: pandas/io/parsers.py (hunk 12 above)
Comment (jreback) [inline pandas/io/parsers.py]: add a note what this is doing

File: doc/source/whatsnew/v0.18.1.txt [hunk 20]
```diff
@@ -211,6 +215,12 @@ Bug Fixes
 
 - Bug in ``value_counts`` when ``normalize=True`` and ``dropna=True`` where nulls still contributed to the normalized count (:issue:`12558`)
 - Bug in ``Panel.fillna()`` ignoring ``inplace=True`` (:issue:`12633`)
+
+- Bug in ``read_csv`` when specifying ``usecols`` and ``parse_dates`` simultaneously with the C engine (:issue:`9755`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.18.1.txt]: and ``names`` right?

Comment (gfyoung) [inline doc/source/whatsnew/v0.18.1.txt]: Yep, fixed.

File: pandas/io/parsers.py (hunk 19 above)
Comment (gfyoung) [inline pandas/io/parsers.py]: Done.

File: pandas/io/parsers.py (hunk 18 above)
Comment (gfyoung) [inline pandas/io/parsers.py]: Right, done.

File: pandas/io/parsers.py (hunk 17 above)
Comment (gfyoung) [inline pandas/io/parsers.py]: Yeah...doesn't look like it.  Removed.

File: doc/source/whatsnew/v0.18.1.txt [hunk 21]
```diff
@@ -211,6 +215,12 @@ Bug Fixes
 
 - Bug in ``value_counts`` when ``normalize=True`` and ``dropna=True`` where nulls still contributed to the normalized count (:issue:`12558`)
 - Bug in ``Panel.fillna()`` ignoring ``inplace=True`` (:issue:`12633`)
+
+- Bug in ``read_csv`` when specifying ``names``, ```usecols``, and ``parse_dates`` simultaneously with the C engine (:issue:`9755`)
+
```
Comment (jreback) [inline doc/source/whatsnew/v0.18.1.txt]: fine for now, but pls don't add extra lines

File: doc/source/io.rst [hunk 22]
```diff
@@ -120,8 +120,8 @@ index_col :  int or sequence or ``False``, default ``None``
   each line, you might consider ``index_col=False`` to force pandas to *not* use
   the first column as the index (row names).
 usecols : array-like, default ``None``
-  Return a subset of the columns. Results in much faster parsing time and lower
-  memory usage
+  Return a subset of the columns. All elements in this array must either be
```
Comment (jreback) [inline doc/source/io.rst]: ideally make these exactly the same (as the doc-string)

File: pandas/io/parsers.py [hunk 23]
```diff
@@ -801,6 +805,16 @@ def _is_index_col(col):
     return col is not None and col is not False
 
 
+def _validate_usecols_arg(usecols):
+    # gh-12678
```
Comment (jreback) [inline pandas/io/parsers.py]: need a 1-liner about what this is validating (as a comment)

File: pandas/io/parsers.py [hunk 24]
```diff
@@ -1133,6 +1147,7 @@ def __init__(self, src, **kwds):
 
         # XXX
         self.usecols = self._reader.usecols
```
Comment (jreback) [inline pandas/io/parsers.py]: better to ``self.usecols = _validate_usecols(self._reader.usecols)``

File: pandas/io/parsers.py [hunk 25]
```diff
@@ -801,6 +805,16 @@ def _is_index_col(col):
...
+def _validate_usecols_arg(usecols):
+    # gh-12678
+    if usecols is not None:
+        usecols_dtype = lib.infer_dtype(usecols)
+        if usecols_dtype not in ('integer', 'string'):
+            raise ValueError(("The elements of 'usecols' "
+                              "must either be all strings "
+                              "or all integers"))
+
+
```
Comment (jreback) [inline pandas/io/parsers.py]: ``return usecols``

File: pandas/io/parsers.py (hunk 12 above)
Comment (jreback) [inline pandas/io/parsers.py]: add ``Parameters / Returns``

File: pandas/io/parsers.py [hunk 26]
```diff
@@ -1472,7 +1494,10 @@ def __init__(self, f, **kwds):
         self.lineterminator = kwds['lineterminator']
         self.quoting = kwds['quoting']
         self.mangle_dupe_cols = kwds.get('mangle_dupe_cols', True)
+
         self.usecols = kwds['usecols']
```
Comment (jreback) [inline pandas/io/parsers.py]: same as above

Comment (gfyoung) [inline pandas/io/parsers.py]: Done.

File: pandas/io/parsers.py (hunk 25 above)
Comment (gfyoung) [inline pandas/io/parsers.py]: Done.

File: pandas/io/parsers.py (hunk 24 above)
Comment (gfyoung) [inline pandas/io/parsers.py]: Done.

File: pandas/io/parsers.py (hunk 23 above)
Comment (gfyoung) [inline pandas/io/parsers.py]: Done.

File: doc/source/io.rst (hunk 22 above)
Comment (gfyoung) [inline doc/source/io.rst]: Yeah...not sure why I didn't copy the rest of the documentation.  Fixed.

File: doc/source/whatsnew/v0.18.1.txt (hunk 21 above)
Comment (gfyoung) [inline doc/source/whatsnew/v0.18.1.txt]: Done.

**Comment by jreback** [comment]
thanks @gfyoung

**Comment by gfyoung** [comment]
@jreback : FYI, you can also cancel both of my builds for this PR on Travis.
